# TeleCo Company: Data Analysis

## Loading Data

In [1]:
import sys
sys.path.insert(0,'../scripts/')

In [2]:
import pandas as pd

In [3]:
from data_loader import load_df_from_csv, load_df_from_excel

In [4]:
fields_description = load_df_from_csv("../data/fields_description.csv")
pd.set_option('display.max_colwidth', None)
fields_description.head()

Fields  \
0  bearer id   
1  Dur. (ms)   
2      Start   
3   Start ms   
4        End   

                                                             Description  
0                                                 xDr session identifier  
1                                      Total Duration of the xDR (in ms)  
2                          Start time of the xDR (first frame timestamp)  
3  Milliseconds offset of start time for the xDR (first frame timestamp)  
4                             End time of the xDR (last frame timestamp)

In [5]:
telData = load_df_from_excel("../data/teleCo_data.xlsx", na_values=['undefined'])
pd.set_option('max_column', None)
telData.head()

Bearer Id               Start  Start ms                 End  \
0  13114483460844900352 2019-04-04 12:01:18     770.0 2019-04-25 14:35:31   
1  13114483482878900224 2019-04-09 13:04:04     235.0 2019-04-25 08:15:48   
2  13114483484080500736 2019-04-09 17:42:11       1.0 2019-04-25 11:58:13   
3  13114483485442799616 2019-04-10 00:31:25     486.0 2019-04-25 07:36:35   
4  13114483499480700928 2019-04-12 20:10:23     565.0 2019-04-25 10:40:32   

   End ms  Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0   662.0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1   606.0  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2   652.0  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3   171.0  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4   954.0  1089009.0  2.082014e+14   3.369979e+10  3.540701e+13   

  Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0   9164566995485190             42.0              5.0   
1            L77566A             65.0              5.0   
2            D42335A              NaN              NaN   
3            T21824A              NaN              NaN   
4            D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0      

### Guidance Sub-tasks

#### Top 10 Handsets used by customers

In [34]:
hand_set = telData['Handset Type'].value_counts()[:10]
hand_set

Huawei B528S-23A                19752
Apple iPhone 6S (A1688)          9419
Apple iPhone 6 (A1586)           9023
Apple iPhone 7 (A1778)           6326
Apple iPhone Se (A1723)          5187
Apple iPhone 8 (A1905)           4993
Apple iPhone Xr (A2105)          4568
Samsung Galaxy S8 (Sm-G950F)     4520
Apple iPhone X (A1901)           3813
Samsung Galaxy A5 Sm-A520F       3724
Name: Handset Type, dtype: int64

#### Top 3 Handset Manufacturers

In [32]:
hand_set_manuf = telData['Handset Manufacturer'].value_counts()[:3]
hand_set_manuf


Apple      59565
Samsung    40839
Huawei     34423
Name: Handset Manufacturer, dtype: int64

#### Top 5 Handset per top 3 Handset Manufacturer

In [115]:
new_df = telData.loc[:,['Handset Type','Handset Manufacturer']]
value = new_df.groupby(['Handset Manufacturer', 'Handset Type']).size()
total_list = pd.Series(dtype='object')
for i in hand_set_manuf.index:
    total_list = total_list.append(value[i])
total_list.sort_values(ascending=False)[:5]

Huawei B528S-23A           19752
Apple iPhone 6S (A1688)     9419
Apple iPhone 6 (A1586)      9023
Apple iPhone 7 (A1778)      6326
Apple iPhone Se (A1723)     5187
dtype: int64

#### Interpretation and Recommendation

The phones widely used in the network are manufactured by the Apple, Samsung and Huawei, contributing for 164,827 users. The most used phone is Hwawei B528S-23A covering 19752 of the phones followed by Apple iPhpone 6S, 6, 7 and Se. The Huawei B528S-23A is favoured by users, so we know that selling similar Huawei phones, make promotions on its newer version if it comes and selling the same phone are profitable. We also can infer that iPhones are the next widely used phones and can implement similar marketing or business ideas to them as well. 

## Variable Identification

For better understanding of the columns, exploring the data description [here](https://docs.google.com/spreadsheets/d/1pcNqeUeIph6xAQzlI54KCvi8HM91SUNeeDbdOq3rvbE/edit?usp=sharing) will help

## Data Understanding

In [6]:
from data_information import DataInfo
explorer = DataInfo(telData)

In [7]:
explorer.get_columns()

Columns Listed in the DataFrame are: 


['Bearer Id',
 'Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Socia

In [8]:
explorer.get_basic_description()

The DataFrame containes 150001 rows and 55 columns.
Current DataFrame Memory Usage:
68672543
Current DataFrame Memory Usage of columns is :
DataFrame Information: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype         
---  ------                                    --------------   -----         
 0   Bearer Id                                 150001 non-null  object        
 1   Start                                     150000 non-null  datetime64[ns]
 2   Start ms                                  150000 non-null  float32       
 3   End                                       150000 non-null  datetime64[ns]
 4   End ms                                    150000 non-null  float32       
 5   Dur. (ms)                                 150000 non-null  float32       
 6   IMSI                                      149431 non-null  float32       
 7   MSISDN/Nu

In [9]:
explorer.get_description()

DataFrame Description: 


Start ms         End ms     Dur. (ms)          IMSI  \
count  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean      499.188263     498.800903  1.046086e+05  2.082016e+14   
std       288.611816     288.097656  8.103762e+04  2.148807e+10   
min         0.000000       0.000000  7.142000e+03  2.040471e+14   
25%       250.000000     251.000000  5.744050e+04  2.082014e+14   
50%       499.000000     500.000000  8.639900e+04  2.082015e+14   
75%       749.000000     750.000000  1.324302e+05  2.082018e+14   
max       999.000000     999.000000  1.859336e+06  2.140743e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847454e+13       109.795708        17.662884   
std     2.447443e+12  2.241637e+13       619.782715        84.793526   
min     3.360100e+10  4.400152e+11         0.000000         0.000000   
25%     3.365129e+10  3.546071e+13        32.000000         2.000000   
50%     3.366371e+10  3.572201e+13        45.000000         5.000000   
75%     3.368349e+10  8.611971e+13        70.000000        15.000000   
max     8.823971e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
count            150000.000000            150000.000000   
mean              13300.045898              1770.428589   
std               23971.878906              4625.355469   
min                   0.000000                 0.000000   
25%                  43.000000                47.000000   
50%                  63.000000                63.000000   
75%               19710.750000              1120.000000   
max              378160.000000             58613.000000   

       TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
count                 6.185500e+04                 5.335200e+04   
mean                  2.080991e+07                 7.596587e+05   
std                   1.825665e+08                 2.645305e+07   
min                   2.000000e+00                 1.000000e+00   
25%                   3.565150e+04                 4.694750e+03   
50%                   5.687300e+05                 2.094950e+04   
75%                   3.768308e+06                 8.402025e+04   
max                   4.294426e+09                 2.908226e+09   

       DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
count        149247.000000                   149247.000000   
mean             92.844757                        3.069355   
std              13.038032                        6.215232   
min               0.000000                        0.000000   
25%              91.000000                        0.000000   
50%             100.000000                        0.000000   
75%             100.000000                        4.000000   
max             100.000000                       93.000000   

       250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
count                  149247.000000       149247.000000        149209.000000   
mean                        1.717341            1.609654            98.530144   
std                         4.159538            4.828890             4.634285   
min                         0.000000            0.000000             0.000000   
25%                         0.000000            0.000000            99.000000   
50%                         0.000000            0.000000           100.000000   
75%                         1.000000            0.000000           100.000000   
max                       100.000000           94.000000           100.000000   

       10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
count                  149209.000000                   149209.000000   
mean                        0.776749                        0.147987   
std                         3.225177                        1.624523   
min                         0.000000          

In [10]:
explorer.get_column_description('Bearer Id')

count     150001
unique    134709
top             
freq         991
Name: Bearer Id, dtype: object

In [11]:
# Data Inforamtion regarding null-values and type based on Columns
explorer.get_memory_usage()

Current DataFrame Memory Usage of columns is :


Index                                           128
Bearer Id                                   1200008
Start                                       1200008
Start ms                                     600004
End                                         1200008
End ms                                       600004
Dur. (ms)                                    600004
IMSI                                         600004
MSISDN/Number                                600004
IMEI                                         600004
Last Location Name                          1200008
Avg RTT DL (ms)                              600004
Avg RTT UL (ms)                              600004
Avg Bearer TP DL (kbps)                      600004
Avg Bearer TP UL (kbps)                      600004
TCP DL Retrans. Vol (Bytes)                  600004
TCP UL Retrans. Vol (Bytes)                  600004
DL TP < 50 Kbps (%)                          600004
50 Kbps < DL TP < 250 Kbps (%)               600004
250 Kbps < D

### Task 1.1

#### Number of xDR Sessions,Session duration, the total download (DL) and upload (UL) data, the total data volume (in Bytes) during this session for each application Aggregated per User('MSISDN/Number')


In [143]:
pd.set_option('display.float_format', '{:.2f}'.format)

In [144]:
aggs_by_col = {'Bearer Id': 'count', 
               'Dur. (ms)': 'sum',
               'Total UL (Bytes)': 'sum',
               'Total DL (Bytes)': 'sum',
               'Social Media DL (Bytes)': 'sum',
               'Social Media UL (Bytes)': 'sum',
               'Google DL (Bytes)': 'sum',
               'Google UL (Bytes)': 'sum',
               'Email DL (Bytes)': 'sum',
               'Email UL (Bytes)': 'sum',
               'Youtube DL (Bytes)': 'sum',
               'Youtube UL (Bytes)': 'sum',
               'Netflix DL (Bytes)': 'sum',
               'Netflix UL (Bytes)': 'sum',
               'Gaming DL (Bytes)': 'sum',
               'Gaming UL (Bytes)': 'sum',
               'Other DL (Bytes)': 'sum',
               'Other UL (Bytes)': 'sum',
               'Gaming DL (Bytes)': 'sum',
               'Gaming UL (Bytes)': 'sum',
               }

ex1 = telData.groupby('MSISDN/Number').agg(aggs_by_col)
ex1

Bearer Id  Dur. (ms)  Total UL (Bytes)  Total DL (Bytes)  \
MSISDN/Number                                                                  
33601001472.00              2  297950.00       72157568.00      963392640.00   
33601003520.00              1  134969.00       39306820.00      556659648.00   
33601007616.00              3   86982.00      114608056.00     1765123584.00   
33601009664.00              2  253983.00       60009248.00      555208000.00   
33601011712.00              3  214759.00      140257136.00      847126272.00   
...                       ...        ...               ...               ...   
33789980672.00              2  210389.00       70831424.00     1023862016.00   
33789997056.00              2  149798.00       62448672.00     1132500224.00   
3197020930048.00            1  877385.00       37295916.00      194828064.00   
337000045477888.00          1  253030.00       56652840.00      539635008.00   
882397138649088.00          1  869844.00       60456048.00       78697600.00   

                    Social Media DL (Bytes)  Social Media UL (Bytes)  \
MSISDN/Number                                                          
33601001472.00                   4805052.00                 87648.00   
33601003520.00                   3148004.00                 47619.00   
33601007616.00                   3112699.00                 80137.00   
33601009664.00                   3787122.00                  5790.00   
33601011712.00                   4447339.00                104980.00   
...                                     ...                      ...   
33789980672.00                   4200896.00                 49416.00   
33789997056.00                    676534.00                122218.00   
3197020930048.00                  668596.00                 46628.00   
337000045477888.00                496337.00                 25229.00   
882397138649088.00               1500145.00                 45943.00   

                    Google DL (Bytes)  Google UL (Bytes)  Email DL (Bytes)  \
MSISDN/Number                                                                
33601001472.00             7534820.00         2189048.00        3666221.00   
33601003520.00             3343483.00           99643.00        2436500.00   
33601007616.00            19666432.00         8511676.00        4426426.00   
33601009664.00            11993496.00         3269638.00        6513834.00   
33601011712.00            15382479.00         6837538.00        2017004.00   
...                               ...                ...               ...   
33789980672.00             5035730.00         5210743.00        4750550.00   
33789997056.00             8306584.00         4654390.00        2994379.00   
3197020930048.00           8572779.00         1865881.00         842279.00   
337000045477888.00         8215537.00         1615080.00        2989663.00   
882397138649088.00         5985089.00         3233558.00        2518425.00   

                    Email UL (Bytes)  Youtube DL (Bytes)  Youtube UL (Bytes)  \
MSISDN/Number                                                                  
33601001472.00             972922.00         20224452.00         13832319.00   
33601003520.00             768880.00          2137272.00         19196298.00   
33601007616.00            1163713.00         14733094.00         33777228.00   
33601009664.00            1334762.00         12169742.00         22177380.00   
33601011712.00            1557518.00          8090922.00         37233152.00   
...                              ...                 ...                 ...   
33789980672.00             564777.00         20353018.00         17659794.00   
33789997056.00             526633.00         13891172.00         32608242.00   
3197020930048.00           678492.00          9839889.00          2120016.00   
337000045477888.00         328919.00         16690728.00         20044212.00   
882397138649088.00         812549.00         18980320.00         21

## Data Cleaning and Manipulation

In [12]:
explorer.get_total_missing_values()

The total number of missing values is 1048375
12.71 % missing values.


1048375

In [13]:
explorer.get_columns_with_missing_values()

['Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Total UL (Bytes)',


In [14]:
explorer.get_column_based_missing_values()

missing_count            type
Start                                                 1  datetime64[ns]
Start ms                                              1         float32
End                                                   1  datetime64[ns]
End ms                                                1         float32
Dur. (ms)                                             1         float32
IMSI                                                570         float32
MSISDN/Number                                      1066         float32
IMEI                                                572         float32
Last Location Name                                 1153          object
Avg RTT DL (ms)                                   27829         float32
Avg RTT UL (ms)                                   27812         float32
Avg Bearer TP DL (kbps)                               1         float32
Avg Bearer TP UL (kbps)                               1         float32
TCP DL Retrans. Vol (Bytes)                       88146         float32
TCP UL Retrans. Vol (Bytes)                       96649         float32
DL TP < 50 Kbps (%)                                 754         float32
50 Kbps < DL TP < 250 Kbps (%)                      754         float32
250 Kbps < DL TP < 1 Mbps (%)                       754         float32
DL TP > 1 Mbps (%)                                  754         float32
UL TP < 10 Kbps (%)                                 792         float32
10 Kbps < UL TP < 50 Kbps (%)                       792         float32
50 Kbps < UL TP < 300 Kbps (%)                      792         float32
UL TP > 300 Kbps (%)                                792         float32
HTTP DL (Bytes)                                   81474         float32
HTTP UL (Bytes)                                   81810         float32
Activity Duration DL (ms)                             1         float32
Activity Duration UL (ms)                             1         float32
Dur. (ms).1                                           1         float32
Handset Manufacturer                               9559          object
Handset Type                                       9559          object
Nb of sec with 125000B < Vol DL                   97538         float32
Nb of sec with 1250B < Vol UL < 6250B             92894         float32
Nb of sec with 31250B < Vol DL < 125000B          93586         float32
Nb of sec with 37500B < Vol UL                   130254         float32
Nb of sec with 6250B < Vol DL < 31250B            88317         float32
Nb of sec with 6250B < Vol UL < 37500B           111843         float32
Nb of sec with Vol DL < 6250B                       755         float32
Nb of sec with Vol UL < 1250B                       793         float32
Total UL (Bytes)                                      1         float32
Total DL (Bytes)                                      1         float32

In [15]:
explorer.get_column_based_missing_percentage()

total_missing_values  \
Bearer Id                                                    0   
Start                                                        1   
Start ms                                                     1   
End                                                          1   
End ms                                                       1   
Dur. (ms)                                                    1   
IMSI                                                       570   
MSISDN/Number                                             1066   
IMEI                                                       572   
Last Location Name                                        1153   
Avg RTT DL (ms)                                          27829   
Avg RTT UL (ms)                                          27812   
Avg Bearer TP DL (kbps)                                      1   
Avg Bearer TP UL (kbps)                                      1   
TCP DL Retrans. Vol (Bytes)                              88146   
TCP UL Retrans. Vol (Bytes)                              96649   
DL TP < 50 Kbps (%)                                        754   
50 Kbps < DL TP < 250 Kbps (%)                             754   
250 Kbps < DL TP < 1 Mbps (%)                              754   
DL TP > 1 Mbps (%)                                         754   
UL TP < 10 Kbps (%)                                        792   
10 Kbps < UL TP < 50 Kbps (%)                              792   
50 Kbps < UL TP < 300 Kbps (%)                             792   
UL TP > 300 Kbps (%)                                       792   
HTTP DL (Bytes)                                          81474   
HTTP UL (Bytes)                                          81810   
Activity Duration DL (ms)                                    1   
Activity Duration UL (ms)                                    1   
Dur. (ms).1                                                  1   
Handset Manufacturer                                      9559   
Handset Type                                              9559   
Nb of sec with 125000B < Vol DL                          97538   
Nb of sec with 1250B < Vol UL < 6250B                    92894   
Nb of sec with 31250B < Vol DL < 125000B                 93586   
Nb of sec with 37500B < Vol UL                          130254   
Nb of sec with 6250B < Vol DL < 31250B                   88317   
Nb of sec with 6250B < Vol UL < 37500B                  111843   
Nb of sec with Vol DL < 6250B                              755   
Nb of sec with Vol UL < 1250B                              793   
Social Media DL (Bytes)                                      0   
Social Media UL (Bytes)                                      0   
Google DL (Bytes)                                            0   
Google UL (Bytes)                                            0   
Email DL (Bytes)                                             0   
Email UL (Bytes)                                             0   
Youtube DL (Bytes)                                           0   
Youtube UL (Bytes)                                           0   
Netflix DL (Bytes)                                           0   
Netflix UL (Bytes)                                           0   
Gaming DL (Bytes)                                            0   
Gaming UL (Bytes)                                            0   
Other DL (Bytes)                                             0   
Other UL (Bytes)                                             0   
Total UL (Bytes)                                             1   
Total DL (Bytes)                                             1   

                                         missing_percentage  
Bearer Id                                             0.0 %  
Start                                                 0.0 %  
Start ms                                              0.0 %  
End                                                   0.0 %  
End ms                             

In [16]:
explorer.get_columns_missing_percentage_greater_than(30.0)

Index(['TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)',
       'HTTP DL (Bytes)', 'HTTP UL (Bytes)', 'Nb of sec with 125000B < Vol DL',
       'Nb of sec with 1250B < Vol UL < 6250B',
       'Nb of sec with 31250B < Vol DL < 125000B',
       'Nb of sec with 37500B < Vol UL',
       'Nb of sec with 6250B < Vol DL < 31250B',
       'Nb of sec with 6250B < Vol UL < 37500B'],
      dtype='object')

In [17]:
from data_cleaner import DataCleaner
cleaner = DataCleaner(explorer.df)
cleaner.df.head()

Bearer Id               Start  Start ms                 End  \
0  13114483460844900352 2019-04-04 12:01:18     770.0 2019-04-25 14:35:31   
1  13114483482878900224 2019-04-09 13:04:04     235.0 2019-04-25 08:15:48   
2  13114483484080500736 2019-04-09 17:42:11       1.0 2019-04-25 11:58:13   
3  13114483485442799616 2019-04-10 00:31:25     486.0 2019-04-25 07:36:35   
4  13114483499480700928 2019-04-12 20:10:23     565.0 2019-04-25 10:40:32   

   End ms  Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0   662.0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1   606.0  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2   652.0  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3   171.0  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4   954.0  1089009.0  2.082014e+14   3.369979e+10  3.540701e+13   

  Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0   9164566995485190             42.0              5.0   
1            L77566A             65.0              5.0   
2            D42335A              NaN              NaN   
3            T21824A              NaN              NaN   
4            D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0      

In [18]:
cleaner.df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype         
---  ------                                    --------------   -----         
 0   Bearer Id                                 150001 non-null  object        
 1   Start                                     150000 non-null  datetime64[ns]
 2   Start ms                                  150000 non-null  float32       
 3   End                                       150000 non-null  datetime64[ns]
 4   End ms                                    150000 non-null  float32       
 5   Dur. (ms)                                 150000 non-null  float32       
 6   IMSI                                      149431 non-null  float32       
 7   MSISDN/Number                             148935 non-null  float32       
 8   IMEI                                      149429 non-null  float32       
 9   Last Location N

In [19]:
cleaner.remove_unwanted_columns(explorer.get_columns_missing_percentage_greater_than(30.0))
cleaner.df.shape

(150001, 45)

In [20]:
cleaner.df.drop([cleaner.df.index[-1]], inplace=True)
cleaner.df.shape

(150000, 45)

In [21]:
explorer = DataInfo(cleaner.df)

In [22]:
explorer.get_columns_with_missing_values()

['IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B']

In [23]:
explorer.get_column_based_missing_values()

missing_count     type
IMSI                                      569  float32
MSISDN/Number                            1065  float32
IMEI                                      571  float32
Last Location Name                       1152   object
Avg RTT DL (ms)                         27828  float32
Avg RTT UL (ms)                         27811  float32
DL TP < 50 Kbps (%)                       753  float32
50 Kbps < DL TP < 250 Kbps (%)            753  float32
250 Kbps < DL TP < 1 Mbps (%)             753  float32
DL TP > 1 Mbps (%)                        753  float32
UL TP < 10 Kbps (%)                       791  float32
10 Kbps < UL TP < 50 Kbps (%)             791  float32
50 Kbps < UL TP < 300 Kbps (%)            791  float32
UL TP > 300 Kbps (%)                      791  float32
Handset Manufacturer                     9558   object
Handset Type                             9558   object
Nb of sec with Vol DL < 6250B             754  float32
Nb of sec with Vol UL < 1250B             792  float32

Check if skew for numeric missing values and if not fill missing values with mean value but if skew use median

In [24]:
cleaner.remove_duplicates()

In [25]:
cleaner.fill_numeric_values(explorer.get_columns_with_missing_values())
cleaner.df.sample(10)

Bearer Id               Start  Start ms  \
110957  13114483567549999104 2019-04-27 19:23:48     846.0   
13602   13042425959082299392 2019-04-24 18:22:04     865.0   
16894   13042425959321200640 2019-04-24 19:19:57     158.0   
73948    7349883254890789888 2019-04-26 08:59:19     771.0   
149468  13042425979312900096 2019-04-29 07:23:28     874.0   
19991   13042425959481300992 2019-04-24 20:03:17     463.0   
123948  13042425975137699840 2019-04-28 05:49:50     345.0   
52963    7349883251244210176 2019-04-25 08:04:48       7.0   
108298   7349883259486400512 2019-04-27 15:02:52     752.0   
111585  13114483567800799232 2019-04-27 20:31:14      29.0   

                       End  End ms  Dur. (ms)          IMSI  MSISDN/Number  \
110957 2019-04-29 06:32:29   580.0   126520.0  2.082017e+14   3.369546e+10   
13602  2019-04-24 23:38:46   317.0    19001.0  2.082010e+14   3.364729e+10   
16894  2019-04-25 02:42:14   767.0    26537.0  2.082014e+14   3.363655e+10   
73948  2019-04-27 04:23:26   995.0    69847.0  2.082009e+14   3.367909e+10   
149468 2019-04-30 07:23:28   856.0    86399.0  2.082014e+14   3.362682e+10   
19991  2019-04-25 01:45:50    37.0    20552.0  2.082019e+14   3.362869e+10   
123948 2019-04-29 05:49:50   297.0    86399.0  2.082014e+14   3.366619e+10   
52963  2019-04-27 06:07:13   170.0   165745.0  2.082018e+14   3.367222e+10   
108298 2019-04-29 09:20:25   330.0   152252.0  2.082019e+14   3.365062e+10   
111585 2019-04-29 07:52:13    57.0   127259.0  2.082009e+14   3.363106e+10   

                IMEI Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
110957  3.566261e+13            L70858B             45.0              5.0   
13602   3.522041e+13            D70228A             57.0             76.0   
16894   3.572291e+13            L72220C            105.0              5.0   
73948   3.573581e+13            L54322C             28.0              1.0   
149468  3.589921e+13            T21316B             38.0             19.0   
19991   3.566111e+13            T23747C            215.0             11.0   
123948  3.533251e+13            L40629B             38.0              5.0   
52963   8.698960e+13            D12723A             14.0              2.0   
108298  3.570631e+13   9164566985578500             45.0              5.0   
111585  3.573691e+13            T79675A             33.0              5.0   

        Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  DL TP < 50 Kbps (%)  \
110957                     43.0                     45.0                100.0   
13602                   44153.0                   2853.0                 88.0   
16894                   53788.0                   1960.0                 86.0   
73948                      39.0                     54.0                100.0   
149468                  14070.0                   1202.0                 96.0   
19991                   15473.0                   1800.0                 41.0   
123948                  47054.0                   5145.0                 53.0   
52963                      44.0                     43.0                100.0   
108298                     59.0                     51.0                100.0   
111585                     10.0                     54.0                100.0   

        50 Kbps < DL TP < 250 Kbps (%)  250 Kbps < DL TP < 1 Mbps (%)  \
110957                             0.0                            0.0   
13602                              1.0                            1.0   
16894                              6.0                            4.0   
73948                              0.0                            0.0   
149468                             2.0                            0.0   
19991                             32.0                           22.0   
123948                            23.0                           13.0   
52963                              0.0                            0.0   
108298                             0.0                            0.0   
111585 

In [26]:
explorer = DataInfo(cleaner.df)
explorer.get_columns_with_missing_values()

['Last Location Name', 'Handset Manufacturer', 'Handset Type']

In [27]:
explorer.df['Last Location Name'].value_counts()

D41377B             80
D17085A             59
D72396C             57
D73604C             55
9164566985995010    55
                    ..
T88953C              1
D64136B              1
D77751A              1
D30008B              1
D10607A              1
Name: Last Location Name, Length: 45547, dtype: int64

In [28]:
explorer.df['Handset Manufacturer'].value_counts()

Apple                                   59565
Samsung                                 40839
Huawei                                  34423
Sony Mobile Communications Ab             980
Wiko Global Sasu                          747
                                        ...  
Shenzhen Huadoo Bright Group Limited        1
Sintave Technology (Group) Co Ltd           1
Ovvi-Cellphone Tech. Inc.                   1
Vastking Technology (Hk) Limited            1
Shenzhen Malata                             1
Name: Handset Manufacturer, Length: 169, dtype: int64

In [29]:
explorer.df['Handset Type'].value_counts()

Huawei B528S-23A                     19752
Apple iPhone 6S (A1688)               9419
Apple iPhone 6 (A1586)                9023
Apple iPhone 7 (A1778)                6326
Apple iPhone Se (A1723)               5187
                                     ...  
Ddm Brands Llc Andy C5V                  1
Blu Products Inc Blu Studio C 5+5        1
Samsung Galaxy Wave 2 (Gt-S8530)         1
Huawei Bln-Al10                          1
Concox Informati. Jh09                   1
Name: Handset Type, Length: 1395, dtype: int64

In [30]:
cleaner.fill_non_numeric_values(explorer.get_columns_with_missing_values(),ffill=True,bfill=True)

TypeError: fill_non_numeric_values() missing 1 required positional argument: 'replaceValues'

In [31]:
telData['Handset Manufacturer'].value_counts()

Apple                                              59565
Samsung                                            40839
Huawei                                             34423
undefined                                           8987
Sony Mobile Communications Ab                        980
                                                   ...  
Sintave Technology (Group) Co Ltd                      1
Shanghai Chengyu Media Technology Co Ltd               1
M S Soha Enterprise                                    1
Essential Products Inc.                                1
Shenzhen Chainway Information Technology Co Ltd        1
Name: Handset Manufacturer, Length: 170, dtype: int64